srcmedia/ts/members-search.ts

```diff
@@ -0,0 +1,91 @@
+import { merge, partition, Observable } from 'rxjs'
+import { pluck, map, withLatestFrom, startWith } from 'rxjs/operators'
+
+import { RxTextInput, RxCheckboxInput } from './lib/input'
+import { RxOutput } from './lib/output'
+import { RxSearchForm } from './lib/form'
+import { RxSelect } from './lib/select'
+import PageControls from './components/PageControls'
+
+document.addEventListener('DOMContentLoaded', () => {
+
+    /* ELEMENTS */
+    const $membersSearchForm = document.getElementById('members-form') as HTMLFormElement
+    const $keywordInput = document.querySelector('input[name=query]') as HTMLInputElement
+    const $hasCardInput = document.querySelector('input[name=has_card]') as HTMLInputElement
+    const $resultsOutput = document.querySelector('output[form=members-form]') as HTMLOutputElement
+    const $totalResultsOutput = document.querySelector('output.total-results') as HTMLOutputElement
+    const $pageSelect = document.querySelector('select[name=page]') as HTMLSelectElement
+    const $sortSelect = document.querySelector('select[name=sort]') as HTMLSelectElement
+    const $pageControls = document.getElementsByClassName('sort-pages')[0] as HTMLElement
+
+    /* COMPONENTS */
+    const membersSearchForm = new RxSearchForm($membersSearchForm)
+    const keywordInput = new RxTextInput($keywordInput)
+    const hasCardInput = new RxCheckboxInput($hasCardInput)
+    const resultsOutput = new RxOutput($resultsOutput)
+    const totalResultsOutput = new RxOutput($totalResultsOutput)
+    const pageSelect = new RxSelect($pageSelect)
+    const sortSelect = new RxSelect($sortSelect)
+    const pageControls = new PageControls($pageControls)
+
+    /* BINDINGS */
+
+    // When a keyword search is active, switch to relevance
+    // Otherwise switch to name
+    // partition() is like filter() but creates two observables: those that
+    // fulfill the condition and those that don't
+    const [keywordOn, keywordOff] = partition(
+        keywordInput.state.pipe(pluck('value')), value => value === '')
+    keywordOn.subscribe(() => sortSelect.update({ value: 'name' }))
+    keywordOff.subscribe(() => sortSelect.update({ value: 'relevance' }))
+
+    // When a user changes the search or sort, go back to page 1
+    merge(
+        keywordInput.state,
+        hasCardInput.state,
+        sortSelect.state,
```

Comment:
i think the criteria for changes that push you back to page 1 is maybe just "every input associated with the form _except_ the pagination itself"? we could move towards a system of automatically detecting these (i think we discussed this briefly since the form has access to all its associated inputs, and PPA does this with `ReactiveForm`) but right now i'm more in the camp of keeping all of the "wiring" outside the components and keeping it explicit. not firmly committed to that though.